src/datatypes/src/types/cast.rs

```diff
@@ -0,0 +1,296 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::data_type::{ConcreteDataType, DataType};
+use crate::error::{self, Error, Result};
+use crate::types::TimeType;
+use crate::value::Value;
+
+/// Cast options for cast functions.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+pub struct CastOption {
+    /// decide how to handle cast failures,
+    /// either return NULL (strict=false) or return ERR (strict=true)
+    pub strict: bool,
+}
+
+impl CastOption {
+    pub fn is_strict(&self) -> bool {
+        self.strict
+    }
+}
+
+/// Cast the value to dest_type with CastOption.
+///
+/// # Arguments
+/// * `src_value` - The value to be casted.
+/// * `dest_type` - The destination type.
+/// * `cast_option` - The CastOption.
+///
+/// # Returns
+/// If success, return the casted value.
+/// If CastOption's strict is true, return an error if the cast fails.
+/// If CastOption's strict is false, return NULL if the cast fails.
+pub fn cast_with_opt(
+    src_value: Value,
+    dest_type: &ConcreteDataType,
+    cast_option: &CastOption,
+) -> Result<Value> {
+    if !can_cast_type(&src_value, dest_type) {
+        if cast_option.strict {
+            return Err(invalid_type_cast(&src_value, dest_type));
+        } else {
+            return Ok(Value::Null);
+        }
+    }
+    let new_value = dest_type.try_cast(src_value.clone());
+    match new_value {
+        Some(v) => Ok(v),
+        None => {
+            if cast_option.strict {
+                Err(invalid_type_cast(&src_value, dest_type))
+            } else {
+                Ok(Value::Null)
+            }
+        }
+    }
+}
+
+// Return true if the src_value can be casted to dest_type,
+// Otherwise, return false.
```

Comment:
- prefer doc comment
- it would better to doc the relation between this fn and `cast_with_opt`. Will `cast_with_opt` fail if this fn returns `true`?

Reply:
- ok, I will rewirte it. 
- yes, Even if `can_cast_type` return true, `cast_with_opt` may still fail when actually casting.